CI broken on main: `AttributeError: module 'headroom.cli' has no attribute 'mcp'` cascade in test_cli/

## Summary

`main`'s CI has been red since #229 merged (2026-04-22 03:14 UTC). All four Python test matrix rows (`test (3.10/3.11/3.12/3.13)`) fail at `tests/test_cli/test_mcp.py` setup and the failure cascades into ~15 other `test_cli/*` tests.

## Reproduction

```
$ pytest tests/test_cli/
...
tests/test_cli/test_mcp.py::TestMCPConfigFunctions::test_get_headroom_command_returns_list PASSED
tests/test_cli/test_mcp.py::TestMCPConfigFunctions::test_load_mcp_config_empty_when_no_file ERROR
...
E   AttributeError: module 'headroom.cli' has no attribute 'mcp'
```

Traceback points at [`tests/test_cli/test_mcp.py:58`](https://github.com/chopratejas/headroom/blob/main/tests/test_cli/test_mcp.py#L58):

```python
with patch("headroom.cli.mcp.MCP_CONFIG_PATH", config_path):
```

## Root cause

[`tests/test_cli/test_wrap_copilot.py:17-23`](https://github.com/chopratejas/headroom/blob/main/tests/test_cli/test_wrap_copilot.py#L17-L23) (added by #229) mutates `sys.modules` at **module import time**:

```python
fake_main_module = types.ModuleType("headroom.cli.main")
fake_main_module.main = click.Group()
sys.modules["headroom.cli.main"] = fake_main_module
sys.modules.pop("headroom.cli", None)
sys.modules.pop("headroom.cli.wrap", None)
wrap_cli = importlib.import_module("headroom.cli.wrap")
```

Effect: when pytest collection imports `test_wrap_copilot.py`, it replaces `sys.modules["headroom.cli.main"]` with a fake `click.Group()` and pops `sys.modules["headroom.cli"]`. The subsequent `importlib.import_module("headroom.cli.wrap")` re-imports `headroom.cli`, which runs `headroom/cli/__init__.py` → `from .main import main` — but `main` is now the fake, so `_register_commands()` never runs. The fresh `headroom.cli` package object has only `.wrap` bound; `.mcp`, `.install`, `.init`, `.evals`, `.learn`, `.memory`, `.perf`, `.proxy`, `.tools` are all missing.

When later tests do `patch("headroom.cli.mcp.MCP_CONFIG_PATH", ...)`, `unittest.mock.patch._dot_lookup` fails at `getattr(headroom.cli, 'mcp')`. The recovery path `__import__("headroom.cli.mcp")` is a no-op because the submodule is still cached in `sys.modules` from the original (now-garbage) package load — `__import__` returns the cached module without re-binding it onto the new package object.

## Failing tests on CI (test (3.10), run `24760244918`)

- `tests/test_cli/test_mcp.py::*` — all `mock_claude_config_path` fixtures ERROR at setup.
- `tests/test_cli/test_install_cli.py::*` — 9 tests.
- `tests/test_cli/test_main_help_version.py::test_subcommand_verbose_flag_still_works`.
- `tests/test_cli/test_wrap_bridge.py::test_wrap_claude_prepare_only_skips_host_binary_lookup`.

## Proposed fix

Eagerly bind subcommand submodules as attributes of the `headroom.cli` package from `headroom/cli/__init__.py` so the attribute lookup survives any sys.modules mutation that re-imports the package through a path other than `main.py`. Defensive — does not touch the test doing the mutation.

PR forthcoming.

--- a/headroom/cli/__init__.py
+++ b/headroom/cli/__init__.py
@@ -1,5 +1,33 @@
-"""Headroom CLI - Command-line interface for memory and proxy management."""
+"""Headroom CLI - Command-line interface for memory and proxy management.
 
+The subcommand submodules are imported eagerly below so they are bound as
+attributes of `headroom.cli`. Click registration happens via side effects in
+`main.py::_register_commands`, but that only binds them to the *main.py*
+module. Tests that do `patch("headroom.cli.<sub>.<attr>")` resolve the target
+by walking attributes on the package object, and that lookup fails when a
+prior test has popped `headroom.cli` from `sys.modules` and re-imported it
+through a path other than `main.py` (e.g. a test that replaces
+`sys.modules["headroom.cli.main"]` with a fake to isolate one subcommand).
+Doing `from . import ...` here means the submodule attribute binding
+survives that kind of sys.modules mutation.
+"""
+
+from . import (  # noqa: F401
+    evals,
+    init,
+    install,
+    learn,
+    mcp,
+    perf,
+    proxy,
+    tools,
+    wrap,
+)
 from .main import main
+
+try:
+    from . import memory  # noqa: F401
+except ImportError:
+    pass
 
 __all__ = ["main"]

--- a/tests/test_cli/test_wrap_copilot.py
+++ b/tests/test_cli/test_wrap_copilot.py
@@ -2,26 +2,15 @@
 
 from __future__ import annotations
 
-import importlib
-import sys
-import types
 from pathlib import Path
 from unittest.mock import patch
 
-import click
 import pytest
 from click.testing import CliRunner
 
+from headroom.cli import wrap as wrap_cli
+from headroom.cli.main import main
 from headroom.copilot_auth import DEFAULT_API_URL
-
-fake_main_module = types.ModuleType("headroom.cli.main")
-fake_main_module.main = click.Group()
-sys.modules["headroom.cli.main"] = fake_main_module
-sys.modules.pop("headroom.cli", None)
-sys.modules.pop("headroom.cli.wrap", None)
-
-wrap_cli = importlib.import_module("headroom.cli.wrap")
-main = fake_main_module.main
 
 
 @pytest.fixture

--- a/tests/test_proxy_copilot_auth_hooks.py
+++ b/tests/test_proxy_copilot_auth_hooks.py
@@ -11,20 +11,20 @@ import pytest
 ROOT = Path(__file__).resolve().parents[1]
 
 
-def _load_handler_module(module_name: str, relative_path: str):
+def _load_handler_module(monkeypatch: pytest.MonkeyPatch, module_name: str, relative_path: str):
     proxy_pkg = types.ModuleType("headroom.proxy")
     proxy_pkg.__path__ = [str(ROOT / "headroom" / "proxy")]
-    sys.modules["headroom.proxy"] = proxy_pkg
+    monkeypatch.setitem(sys.modules, "headroom.proxy", proxy_pkg)
 
     handlers_pkg = types.ModuleType("headroom.proxy.handlers")
     handlers_pkg.__path__ = [str(ROOT / "headroom" / "proxy" / "handlers")]
-    sys.modules["headroom.proxy.handlers"] = handlers_pkg
+    monkeypatch.setitem(sys.modules, "headroom.proxy.handlers", handlers_pkg)
 
     httpx_mod = types.ModuleType("httpx")
     httpx_mod.ConnectError = type("ConnectError", (Exception,), {})
     httpx_mod.ConnectTimeout = type("ConnectTimeout", (Exception,), {})
     httpx_mod.PoolTimeout = type("PoolTimeout", (Exception,), {})
-    sys.modules["httpx"] = httpx_mod
+    monkeypatch.setitem(sys.modules, "httpx", httpx_mod)
 
     responses_mod = types.ModuleType("fastapi.responses")
 
@@ -40,12 +40,12 @@ def _load_handler_module(module_name: str, relative_path: str):
 
     responses_mod.Response = Response
     responses_mod.StreamingResponse = StreamingResponse
-    sys.modules["fastapi.responses"] = responses_mod
+    monkeypatch.setitem(sys.modules, "fastapi.responses", responses_mod)
 
     spec = importlib.util.spec_from_file_location(module_name, ROOT / relative_path)
     assert spec is not None and spec.loader is not None
     module = importlib.util.module_from_spec(spec)
-    sys.modules[module_name] = module
+    monkeypatch.setitem(sys.modules, module_name, module)
     spec.loader.exec_module(module)
     return module
 
@@ -53,6 +53,7 @@ def _load_handler_module(module_name: str, relative_path: str):
 @pytest.mark.asyncio
 async def test_openai_passthrough_applies_copilot_auth(monkeypatch: pytest.MonkeyPatch) -> None:
     openai_mod = _load_handler_module(
+        monkeypatch,
         "tests.headroom_proxy_handlers_openai",
         "headroom/proxy/handlers/openai.py",
     )
@@ -110,6 +111,7 @@ async def test_openai_passthrough_applies_copilot_auth(monkeypatch: pytest.Monke
 @pytest.mark.asyncio
 async def test_streaming_response_applies_copilot_auth(monkeypatch: pytest.MonkeyPatch) -> None:
     streaming_mod = _load_handler_module(
+        monkeypatch,
         "tests.headroom_proxy_handlers_streaming",
         "headroom/proxy/handlers/streaming.py",
     )

--- a/tests/test_release_version.py
+++ b/tests/test_release_version.py
@@ -14,6 +14,7 @@ from headroom.release_version import (
     compute_release_version,
     determine_bump_level,
     find_latest_release_tag,
+    get_canonical_version,
     list_release_commits,
     normalize_release_tag,
     parse_release_tag,
@@ -177,7 +178,7 @@ def test_release_version_script_runs_directly_without_importing_headroom_package
     assert output_path.read_text(encoding="utf-8").splitlines() == [
         "version=0.6.0",
         "npm_version=0.6.0",
-        "canonical=0.5.25",
+        f"canonical={get_canonical_version(ROOT)}",
         "height=0",
         "bump=manual",
         "previous_tag=",

--- a/tests/test_release_workflows.py
+++ b/tests/test_release_workflows.py
@@ -53,6 +53,8 @@ def test_macos_native_wrapper_dependency_install_retries_pypi_downloads() -> Non
     content = (ROOT / ".github" / "workflows" / "ci.yml").read_text(encoding="utf-8")
 
     assert "python -m pip install --retries 10 --timeout 60 pytest" in content
+
+
 def test_ci_commitlint_skips_default_github_merge_commits() -> None:
     content = (ROOT / ".github" / "workflows" / "ci.yml").read_text(encoding="utf-8")
 

